### PR TITLE
ForceJump mode for NoFall

### PR DESCRIPTION
**Pull request by SmugNugg**
I implemented a ForceJump mode for NoFall.
This jumps just before the player will land.

### CODE REVIEW

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinLivingEntity.java [hunk 1]
```diff
@@ -139,13 +144,17 @@ private Vec3d hookFixRotation(Vec3d instance, double x, double y, double z) {
...
     @Inject(method = "pushAwayFrom", at = @At("HEAD"), cancellable = true)
-    private void hookNoPush(CallbackInfo callbackInfo) {
+    private void hookNoPush(Entity entity, CallbackInfo ci) {
         if (ModuleNoPush.INSTANCE.getEnabled()) {
-            callbackInfo.cancel();
+            // Checking if the entity is a playerEntity or a livingEntity depending if the playerOnly is toggled or not.
+            if ((entity instanceof PlayerEntity && ModuleNoPush.INSTANCE.getPlayerOnly()) ||
+                    (entity instanceof LivingEntity && !ModuleNoPush.INSTANCE.getPlayerOnly())) {
+                ci.cancel();
+            }
```
Comment (superblaubeere27) [inline src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinLivingEntity.java]: This should be it's own PR

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/ModuleNoPush.kt [hunk 2]
```diff
@@ -18,12 +18,21 @@
...
 
 /**
  * NoPush module
  *
  * Disables pushing from other players.
  */
-object ModuleNoPush : Module("NoPush", Category.MOVEMENT)
+object ModuleNoPush : Module("NoPush", Category.MOVEMENT) {
+    public var playerOnly by boolean("Player Only", true)
+}
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/ModuleNoPush.kt]: This too

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import net.minecraft.util.math.BlockPos
+import net.minecraft.util.math.Vec3d
+import kotlin.math.floor
+
+/**
+ * NoFallForceJump mode for the NoFall module.
+ * This mode forces the player to jump just when his about to land,
+ * preventing fall damage.
+ */
+internal object NoFallForceJump : Choice("Force Jump") {
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt]: Spaces are added automatically by the frontend `ForceJump` -> `Force Jump`

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    override val parent: ChoiceConfigurable<*>
+        get() = ModuleNoFall.modes
+
+    private val blockDistance by int("Block Distance", 1, 0..5)
+    private val fallDistance by float("Fall Distance", 3.35f, 3.35f..10.0f)
+    private val jumpHeight by float("Jump Height", 0.42f, 0.1f..0.42f)
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt]: s.a.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+     * This is done by checking if the player's fall distance is higher than the specific (fallDistance)
+     * and if the player is above a nonair block by the specific block distance.
+     */
+    val packetHandler = handler<PacketEvent> {
+        val packet = it.packet
+        if (packet is PlayerMoveC2SPacket && player.fallDistance > fallDistance) {
+            val blockBelowPos = calculateBlockBelowPosition()
+            val blockBelow: BlockState? = getBlockStateAtPosition(blockBelowPos)
+
+            if (!jumpTriggered && blockBelow != null && !blockBelow.isAir) {
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallForceJump.kt]: It would be better (and simpler) if you did this check with `FallingPlayer` (see https://github.com/CCBlueX/LiquidBounce/blob/nextgen/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/NoFallMLG.kt#L178)

**Review by superblaubeere27 — CHANGES_REQUESTED** [review]
Good idea, good work. I have some suggestions though

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/ModuleNoPush.kt (hunk 2 above)
Comment (SmugNugg) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/movement/ModuleNoPush.kt]: > This too

I know i was lost on how to make new pull requests

File: src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinLivingEntity.java (hunk 1 above)
Comment (SmugNugg) [reply src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinLivingEntity.java]: Yes i already made a new pull request for it. 
The reason why this happened is because i didn't know how to make a new pull request sadly 😭 
But I've figured it out.

> This should be it's own PR

**Review by SmugNugg — COMMENTED** [review]
made it better by making request changes

**Review by 1zun4 — APPROVED** [review]

**Comment by 1zun4** [comment]
@SmugNugg I made some changes to your PR because the original code not only did not compile, it did not work properly. I have now fixed this. Thanks for your PR and please try to use this as a learning approach :)

**Review by 1zun4 — APPROVED** [review]
yeah i approve my own changes
